### PR TITLE
cosine embedding loss function in pytorch frontend

### PR DESCRIPTION
**Pull request by satani99**

### CODE REVIEW

**Comment by satani99** [comment]
Issue #7199

**Comment by satani99** [comment]
Close #7199

**Comment by satani99** [comment]
hey @Ishticode check this code

**Comment by satani99** [comment]
I think nightly tests are failing in all PyTorch loss functions

**Review by Ishticode — CHANGES_REQUESTED** [review]
Hi @satani99 
Thanks a lot for the PR. 🙂.
There was a CI issue which failed to instantiate the tests properly at the time. Could you kindly resolve the merge conflicts and this should trigger the tests to be re-run. You can request a review once updated. Thanks. Let me know if you have any query. You can request a review once updated.

**Comment by satani99** [comment]
I think front-end tests run correctly. Ready for review.

File: ivy/functional/frontends/torch/nn/functional/loss_functions.py [hunk 1]
```diff
@@ -191,3 +192,23 @@ def nll_loss(
...
+    input2,
+    target,
+    margin=0.0,
+    size_average=None,
+    reduce=None,
+    reduction="mean",
+):
+    similarity = distance_functions.cosine_similarity(input1, input2)
+    if target == 1:
+        loss = 1 - similarity
```
Comment (Ishticode) [inline ivy/functional/frontends/torch/nn/functional/loss_functions.py:209]: This line throws an error as `TypeError: unsupported operand type(s) for -: 'int' and 'Tensor'`. Perhaps using a debugger to see why the operation isn't supported. Normally an operation as such should be valid. Feel free to ask anything and please request when my review when fixed.

**Review by Ishticode — CHANGES_REQUESTED** [review]

**Review by Ishticode — CHANGES_REQUESTED** [review]
Hi. I see that the whole of the files `linalg.py` and `test_linalg` are deleted. Could you kindly undo that change and request a review.🙂

**Review by Ishticode — CHANGES_REQUESTED** [review]
Hi @satani99 
Thanks for the review request. 
The test `test_torch_cosine_embedding_loss` is failing with error `IndexError: Too many indices for array: 1 non-None/Ellipsis indices for dim 0`. You can see this by clicking on Frontend Pytorch test run "details" down below. Could you kindly fix this and also resolve the conflict. Feel free to follow or ask a question if this doesn't make sense and feel free to request a review when you its ready to be reviewed again. Thank you @satani99

**Review by Ishticode — CHANGES_REQUESTED** [review]
@satani99 
Somehow your last changes ended up removing the import for distance functions and now the test throws error for undefined `distance_function` call. Do you have a setup where u can check for tests. I would advice setting up test running or if not then perhaps checking out the CI logs from below for `test-frontend-torch / run-nightly-tests (pull_request)` 

Minor mistakes like this will take more time of yours than they should. The progress would be quicker if such errors have been taken care of and the tests have been checked out by yourself. 

However, I am totally happy to review whenever you want. This is just a tip to speed up your PR's progress. Hope it makes sense.

Meanwhile could you recover the `distance_function` import and fix any test error pointed to earlier and feel free to ask anything. 🙂

**Comment by satani99** [comment]
How do I test on my local computer?

**Review by Ishticode — COMMENTED** [review]
Hi You can setup with using this [setting up instructions](https://lets-unify.ai/ivy/contributing/setting_up.html) or try one of the relevant vidz from [this youtube playlist](https://www.youtube.com/playlist?list=PLwNuX3xB_tv-7NexAEzC2dCbgj5ojr-0X).
Let me know if anything doesn't quite make sense.  🙂

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Review by Ishticode — CHANGES_REQUESTED** [review]
Hi @satani99 
I am not sure how I missed leaving the review. I remember writing it two days ago but probably did not submit. 
Apologies for the delay anyhow.

Did you manage to setup a system.
The tests still fail for number of reasons.
The main reason is that the `target` argument needs more care. 

Pytorch expects targets to be tensors so we should also consider target to be tensor and treat it like tensors... They should be generated as tensors in the test and not as integers of 1 or -1.

Additionally, the case where target is just provided as a value other than 1 or -1, we should be catching it within the function and raising an exception. Otherwise the loss variable will be undefined... unless you want to define loss = None beforehand.
Please feel free to try these and ask any questions.

**Comment by ivy-seed** [comment]
This PR has been closed because it has been marked as stale for more than 7 days with no activity.

**Pull request closed without merging**